File: vectra/VECTRA_2025-12-18/cpp/src/artifact.cpp
/**
 * @file artifact.cpp
 * @brief Artifact serialization implementation.
 */

#include "vectra/vectra.hpp"

#include <cstring>

namespace vectra {

// Simple binary serialization (production would use proper format)
std::vector<uint8_t> Artifact::to_bytes() const {
    std::vector<uint8_t> result;

    // Version marker
    uint64_t version = VERSION_ID;
    for (int i = 0; i < 8; ++i) {
        result.push_back(static_cast<uint8_t>(version >> (i * 8)));
    }

    // Generator base length and data
    uint32_t base_len = static_cast<uint32_t>(generator.base.size());
    for (int i = 0; i < 4; ++i) {
        result.push_back(static_cast<uint8_t>(base_len >> (i * 8)));
    }
    result.insert(result.end(), generator.base.begin(), generator.base.end());

    // Repetition spec
    for (int i = 0; i < 4; ++i) {
        result.push_back(static_cast<uint8_t>(generator.repetition.count >> (i * 8)));
    }
    for (int i = 0; i < 4; ++i) {
        result.push_back(static_cast<uint8_t>(generator.repetition.stride >> (i * 8)));
    }

    // Residual segments count
    uint32_t seg_count = static_cast<uint32_t>(residual.segments.size());
    for (int i = 0; i < 4; ++i) {
        result.push_back(static_cast<uint8_t>(seg_count >> (i * 8)));
    }

    // Each segment
    for (const auto& seg : residual.segments) {
        // Range
        uint64_t start = seg.range.start;
        uint64_t end = seg.range.end;
        for (int i = 0; i < 8; ++i) {
            result.push_back(static_cast<uint8_t>(start >> (i * 8)));
        }
        for (int i = 0; i < 8; ++i) {
            result.push_back(static_cast<uint8_t>(end >> (i * 8)));
        }

        // Delta length and data
        uint32_t delta_len = static_cast<uint32_t>(seg.delta.size());
        for (int i = 0; i < 4; ++i) {
            result.push_back(static_cast<uint8_t>(delta_len >> (i * 8)));
        }
        result.insert(result.end(), seg.delta.begin(), seg.delta.end());
    }

    // Constraints
    uint64_t output_len = constraints.output_length;
    for (int i = 0; i < 8; ++i) {
        result.push_back(static_cast<uint8_t>(output_len >> (i * 8)));
    }
    result.insert(result.end(), constraints.output_hash.begin(), constraints.output_hash.end());

    // Integrity
    result.insert(result.end(), integrity.payload_hash.begin(), integrity.payload_hash.end());
    result.insert(result.end(), integrity.artifact_hash.begin(), integrity.artifact_hash.end());
    for (int i = 0; i < 8; ++i) {
        result.push_back(static_cast<uint8_t>(integrity.version >> (i * 8)));
    }
    for (int i = 0; i < 8; ++i) {
        result.push_back(static_cast<uint8_t>(integrity.encoded_at >> (i * 8)));
    }

    return result;
}

Artifact Artifact::from_bytes(const std::vector<uint8_t>& data) {
    Artifact artifact;
    size_t pos = 0;

    auto read_u32 = [&]() -> uint32_t {
        uint32_t val = 0;
        for (int i = 0; i < 4 && pos < data.size(); ++i, ++pos) {
            val |= static_cast<uint32_t>(data[pos]) << (i * 8);
        }
        return val;
    };

    auto read_u64 = [&]() -> uint64_t {
        uint64_t val = 0;
        for (int i = 0; i < 8 && pos < data.size(); ++i, ++pos) {
            val |= static_cast<uint64_t>(data[pos]) << (i * 8);
        }
        return val;
    };

    auto read_bytes = [&](size_t len) -> std::vector<uint8_t> {
        std::vector<uint8_t> result;
        for (size_t i = 0; i < len && pos < data.size(); ++i, ++pos) {
            result.push_back(data[pos]);
        }
        return result;
    };

    auto read_hash = [&]() -> Hash256 {
        Hash256 hash{};
        for (size_t i = 0; i < 32 && pos < data.size(); ++i, ++pos) {
            hash[i] = data[pos];
        }
        return hash;
    };

    // Version
    read_u64();  // Skip version marker

    // Generator
    uint32_t base_len = read_u32();
    artifact.generator.base = read_bytes(base_len);
    artifact.generator.repetition.count = read_u32();
    artifact.generator.repetition.stride = read_u32();

    // Residual segments
    uint32_t seg_count = read_u32();
    for (uint32_t i = 0; i < seg_count; ++i) {
        ResidualSegment seg;
        seg.range.start = read_u64();
        seg.range.end = read_u64();
        uint32_t delta_len = read_u32();
        seg.delta = read_bytes(delta_len);
        artifact.residual.segments.push_back(std::move(seg));
    }

    // Constraints
    artifact.constraints.output_length = read_u64();
    artifact.constraints.output_hash = read_hash();

    // Integrity
    artifact.integrity.payload_hash = read_hash();
    artifact.integrity.artifact_hash = read_hash();
    artifact.integrity.version = read_u64();
    artifact.integrity.encoded_at = read_u64();

    // Predictor state (default)
    artifact.predictor_state.version = VERSION_ID;

    return artifact;
}

}  // namespace vectra

File: vectra/vectra/src/artifact.rs
//! Author / Inventor: Katta Naga Sri Ganesh
//! Organization: SYNTRIASS Labs Private Limited
//! Copyright © 2025 SYNTRIASS Labs Private Limited

//! Artifact construction (TDF — VECTRA Data Format)
//!
//! Implements spec §7: Self-describing artifact assembly.
//!
//! The artifact contains everything needed for reconstruction:
//! - G, Φ: Structural generators (from FEE)
//! - Θ: Predictor state (from SPE)
//! - Δ: Bounded residual (approved by EBTA)
//! - C: Reconstruction constraints
//! - I: Integrity metadata

use crate::error::EncodeError;
use crate::integrity::{generate_integrity_metadata, generate_reconstruction_constraints};
use crate::types::{
    Artifact, Generator, MappingSet, Payload, PredictorState, Residual,
};

/// Build complete artifact from components.
///
/// # Preconditions
///
/// - EBTA has validated the residual (entropy within bounds)
/// - All components are deterministically generated
///
/// # Postconditions
///
/// - Artifact is self-describing (contains all reconstruction info)
/// - Artifact is self-verifiable (contains integrity metadata)
/// - Artifact can be decoded without external context
pub fn build_artifact(
    payload: &Payload,
    generator: Generator,
    mappings: MappingSet,
    predictor_state: PredictorState,
    residual: Residual,
) -> Result<Artifact, EncodeError> {
    // Generate reconstruction constraints
    let constraints = generate_reconstruction_constraints(payload);

    // Generate integrity metadata
    let integrity = generate_integrity_metadata(
        payload,
        &generator,
        &mappings,
        &predictor_state,
        &residual,
    );

    Ok(Artifact {
        generator,
        mappings,
        predictor_state,
        residual,
        constraints,
        integrity,
    })
}

/// Estimate artifact size in bytes.
///
/// Useful for deciding whether encoding provides benefit.
pub fn estimate_artifact_size(artifact: &Artifact) -> usize {
    let mut size = 0;

    // Generator
    size += artifact.generator.base.len();
    size += 8; // repetition spec

    // Mappings (estimate)
    size += artifact.mappings.mappings.len() * 24;

    // Predictor state (estimate)
    size += 64;

    // Residual
    for segment in &artifact.residual.segments {
        size += segment.delta.len();
        size += 16; // range metadata
    }

    // Constraints
    size += 8 + 32; // length + hash

    // Integrity
    size += 32 + 32 + 8 + 8; // hashes + version + timestamp

    size
}

/// Check if encoding provides size benefit.
///
/// Returns true if artifact is smaller than original payload.
pub fn is_encoding_beneficial(payload: &Payload, artifact: &Artifact) -> bool {
    let artifact_size = estimate_artifact_size(artifact);
    artifact_size < payload.len()
}

/// Compute compression ratio.
///
/// Returns original_size / artifact_size.
/// Values > 1.0 indicate compression benefit.
pub fn compression_ratio(payload: &Payload, artifact: &Artifact) -> f64 {
    let artifact_size = estimate_artifact_size(artifact);
    if artifact_size == 0 {
        return 1.0;
    }
    payload.len() as f64 / artifact_size as f64
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{
        ByteRange, PredictorParameters, RepetitionSpec, ResidualSegment, VERSION_ID,
    };

    fn create_test_payload() -> Payload {
        Payload::new(vec![0xDE, 0xAD, 0xBE, 0xEF, 0xCA, 0xFE, 0xBA, 0xBE])
    }

    fn create_test_components() -> (Generator, MappingSet, PredictorState, Residual) {
        let generator = Generator {
            base: vec![0xDE, 0xAD],
            repetition: RepetitionSpec { count: 4, stride: 2, start_offset: 0, byte_ranges: vec![] },
        };

        let mappings = MappingSet { mappings: vec![] };

        let predictor_state = PredictorState {
            version: VERSION_ID,
            parameters: PredictorParameters::default(),
        };

        let residual = Residual {
            segments: vec![ResidualSegment {
                range: ByteRange { start: 0, end: 8 },
                delta: vec![0x00; 8],
                semantic_type: crate::types::SemanticType::Opaque,
            }],
        };

        (generator, mappings, predictor_state, residual)
    }

    #[test]
    fn test_build_artifact() {
        let payload = create_test_payload();
        let (generator, mappings, predictor_state, residual) = create_test_components();

        let artifact =
            build_artifact(&payload, generator, mappings, predictor_state, residual).unwrap();

        // Verify structure
        assert_eq!(artifact.generator.base, vec![0xDE, 0xAD]);
        assert_eq!(artifact.integrity.version, VERSION_ID);
        assert_eq!(artifact.constraints.output_length, 8);
    }

    #[test]
    fn test_build_artifact_determinism() {
        let payload = create_test_payload();

        let (g1, m1, p1, r1) = create_test_components();
        let (g2, m2, p2, r2) = create_test_components();

        let artifact1 = build_artifact(&payload, g1, m1, p1, r1).unwrap();
        let artifact2 = build_artifact(&payload, g2, m2, p2, r2).unwrap();

        // Core components must match
        assert_eq!(artifact1.generator.base, artifact2.generator.base);
        assert_eq!(artifact1.constraints.output_hash, artifact2.constraints.output_hash);
        assert_eq!(artifact1.integrity.payload_hash, artifact2.integrity.payload_hash);
    }

    #[test]
    fn test_estimate_artifact_size() {
        let payload = create_test_payload();
        let (generator, mappings, predictor_state, residual) = create_test_components();

        let artifact =
            build_artifact(&payload, generator, mappings, predictor_state, residual).unwrap();

        let size = estimate_artifact_size(&artifact);
        assert!(size > 0);
    }

    #[test]
    fn test_compression_ratio() {
        // Large payload with small residual should have good ratio
        let payload = Payload::new(vec![0xAA; 1000]);

        let generator = Generator {
            base: vec![0xAA],
            repetition: RepetitionSpec {
                count: 1000,
                stride: 1,
                start_offset: 0,
                byte_ranges: vec![],
            },
        };
        let mappings = MappingSet { mappings: vec![] };
        let predictor_state = PredictorState {
            version: VERSION_ID,
            parameters: PredictorParameters::default(),
        };
        let residual = Residual {
            segments: vec![ResidualSegment {
                range: ByteRange { start: 0, end: 1000 },
                delta: vec![0x00; 10], // Very small residual
                semantic_type: crate::types::SemanticType::Opaque,
            }],
        };

        let artifact =
            build_artifact(&payload, generator, mappings, predictor_state, residual).unwrap();

        let ratio = compression_ratio(&payload, &artifact);
        assert!(ratio > 1.0, "Should have compression benefit");
    }
}
